src/enzyme_ad/jax/Passes/AffineToStableHLORaising.cpp

```diff
@@ -634,6 +792,180 @@ tryRaisingOpToStableHLO(Operation *op, IRMapping &mapping, OpBuilder &builder,
     return success();
   }
 
+  // Affine if (only pure ops with yield is currently supported)
+  if (auto ifOp = dyn_cast<affine::AffineIfOp>(op)) {
+    if (!ifOp.hasElse() || ifOp->getNumResults() == 0 ||
+        llvm::any_of(*ifOp.getThenBlock(),
+                     [](Operation &op) { return !mlir::isPure(&op); }) ||
+        llvm::any_of(*ifOp.getElseBlock(),
+                     [](Operation &op) { return !mlir::isPure(&op); })) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "cannot raise if yet (non-pure or yielded values): " << *op
+                 << "\n");
+      return failure();
+    }
+
+    auto is = ifOp.getIntegerSet();
+    if (is.getNumSymbols() != 0) {
+      LLVM_DEBUG(llvm::dbgs() << "cannot raise integer set with symbols yet\n");
+      return failure(); // TODO
+    }
+
+    Value cond = nullptr;
+    affine::AffineValueMap map(AffineMap::get(ifOp.getContext()), {});
+    for (auto [constraint, eq] :
+         llvm::zip_equal(is.getConstraints(), is.getEqFlags())) {
+      auto [expandedExpr, outputMap] = expandAffineExpr(
+          builder, ifOp.getLoc(), constraint, ifOp.getOperands(), mapping);
+      Value zero = builder.create<stablehlo::ConstantOp>(
+          ifOp.getLoc(), expandedExpr.getType().cast<ShapedType>(),
+          SplatElementsAttr::get(
+              expandedExpr.getType().cast<ShapedType>(),
+              ArrayRef<Attribute>(IntegerAttr::get(builder.getI64Type(), 0))));
+      Value newCond = builder.create<stablehlo::CompareOp>(
+          ifOp.getLoc(), expandedExpr, zero,
+          eq ? stablehlo::ComparisonDirection::EQ
+             : stablehlo::ComparisonDirection::GE);
+      if (cond) {
+        map = alignMemoryAccess(cond, map, newCond, outputMap, builder);
+        cond = builder.create<stablehlo::AndOp>(ifOp.getLoc(), cond, newCond);
+      } else {
+        cond = newCond;
+        map = outputMap;
+      }
+    }
+
+    for (auto &innerOp : ifOp.getThenBlock()->without_terminator()) {
+      if (tryRaisingOpToStableHLO(&innerOp, mapping, builder, maps).failed())
```

Comment:
we need an extra check here to confirm that the inner op is speculatable [i.e. won't go out of bounds/etc]